dataset/src/main/java/org/keycloak/benchmark/dataset/ExecutorHelper.java

```diff
@@ -50,29 +50,35 @@ public ExecutorHelper(int threadCount, KeycloakSessionFactory sessionFactory, Da
     }
 
     public void addTask(Runnable task) {
-        Future f = executor.submit(task);
+        Future<?> f = executor.submit(task);
         futures.add(f);
     }
 
     public void addTaskRunningInTransaction(KeycloakSessionTask sessionTask) {
-        Future f = executor.submit(() -> KeycloakModelUtils.runJobInTransactionWithTimeout(sessionFactory, sessionTask, config.getTransactionTimeoutInSeconds()));
+        Future<?> f = executor.submit(() -> KeycloakModelUtils.runJobInTransactionWithTimeout(sessionFactory, sessionTask, config.getTransactionTimeoutInSeconds()));
         futures.add(f);
     }
 
 
-    public void waitForAllToFinish() throws ExecutionException, InterruptedException {
-        logger.info("Waiting for tasks to complete");
-
-        for (Optional<Future> runningTask = getFirstRunningTask(); runningTask.isPresent(); runningTask = getFirstRunningTask()) {
-            runningTask.get().get(); // wait for task to complete
+    public void waitForAllToFinish() {
+        logger.info("Waiting for tasks to complete successfully");
+        Collection<Throwable> failures = new ConcurrentLinkedQueue<>();
+        while (!futures.isEmpty()) {
+            // some tasks will spawn other tasks, so expect the futures queue to be modified concurrently.
+            try {
+                futures.poll().get();
+            } catch (Throwable e) {
+                failures.add(e);
+            }
+        }
+        if (!failures.isEmpty()) {
+            RuntimeException ex = new RuntimeException("Some futures failed");
```

Comment:
would it be possible to log more information about the type of failure which could help with the debug or would it log anyways ?

Reply:
The code creates a Runtime Exception which then has a list of one or more suppressed exceptions. All of the suppressed exceptions will be logged to together with the message "Some futures failed".

The the following line for where the suppressed exceptions are added.